### PR TITLE
Use Argon2i for wallet encryption PBKDF

### LINKED ISSUE #1207: Change wallet encryption PBKDF to Argon2
The PBKDF used to derive wallet encryption is currently just an iterated salted SHA-512 hash, documented at https://www.openssl.org/docs/manmaster/crypto/EVP_BytesToKey.html . The parameters are set in ``CCrypter::SetKeyFromPassphrase`` in ``src/wallet/crypter.cpp``.

Argon2 is a much better PBKDF in terms of dictionary attack resistance. It's [implemented by libsodium](https://download.libsodium.org/libsodium/content/password_hashing/the_argon2i_function.html) so would not add to our dependencies.

### PR DESCRIPTION
**Pull request by str4d**
Closes #1207.

### CODE REVIEW

File: src/wallet/crypter.h [hunk 1]
```diff
@@ -58,11 +59,11 @@ class CMasterKey
...
     {
-        // 25000 rounds is just under 0.1 seconds on a 1.86 GHz Pentium M
-        // ie slightly lower than the lowest hardware we need bother supporting
-        nDeriveIterations = 25000;
         nDerivationMethod = 0;
-        vchOtherDerivationParameters = std::vector<unsigned char>(0);
+        nDeriveIterations = crypto_pwhash_OPSLIMIT_INTERACTIVE;
+        size_t memlimit = crypto_pwhash_MEMLIMIT_INTERACTIVE;
+        vchOtherDerivationParameters = std::vector<unsigned char>(sizeof(size_t));
+        ::memcpy(vchOtherDerivationParameters.data(), &memlimit, sizeof(memlimit));
```
Comment (str4d) [inline src/wallet/crypter.h]: This is something I'm not certain on. AIUI `size_t` is implementation-specific, but libsodium expects it as part of its API. Perhaps instead we should serialize the value in `vchOtherDerivationParameters` as a Bitcoin `compactSize` instead?

Comment (ebfull) [inline src/wallet/crypter.h]: Yes, you should serialize with compact size. Even the endianness here is implementation specific.

**Comment by str4d** [comment]
@ebfull comment addressed.

**Comment by str4d** [comment]
Comments from @daira that belong in this PR:

- https://github.com/zcash/zcash/pull/1984#discussion_r97129255
- https://github.com/zcash/zcash/pull/1984#discussion_r97131037

**Comment by indolering** [comment]
The ideal implementation would use a more conservative iteration count for the final value but cache an intermediate hash that is disassociated from the wallet file and optionally removed on logout.  This drastically improves security without forcing users to wait 5 seconds every-time they move funds.

Sadly, it doesn't look like libsodium has that functionality built in and would require manual work on your part.  I should probably talk to @ryancdotorg about publishing a best-practices document.

**Comment by str4d** [comment]
This is obsolete; we will likely be switching to Argon2id instead, but using Rust dependencies instead of `libsodium`.

**Pull request closed without merging**